memote/support/biomass.py

```diff
@@ -176,14 +192,83 @@ def find_direct_metabolites(model, reaction):
+    if solution.objective_value == 0:
+        LOGGER.error("Failed to generate a non-zero objective value with "
+                     "flux balance analysis.")
+        raise ValueError("The flux balance analysis on this model returned an "
+                         "objective value of zero. Make sure the model can "
+                         "grow! Check if the constraints are not too strict!")
+
+    tra_bou_bio_fluxes = solution.fluxes[get_ids(rxns_of_interest)]
+    flux_sum = pd.DataFrame(index=tra_bou_bio_mets, columns=["sum"], data=0).T
+
+    # The following is to detect false positives.
+    # As mentioned previously, these false positives exists in the "e"
+    # compartment with flux from the "c" compartment and are part of the
+    # biomass reaction(s). It sums fluxes positively or negatively depending
+    # on if direct metabolites in the "e" compartment are defined as reactants
+    # or products in various reactions.
+    for met in tra_bou_bio_mets:
+        for rxn in met.reactions:
+            not_biomass_rxn = rxn not in biomass_rxns
+            extracellular = met.compartment == ext_space
+            # if the metabolite is in the "e" compartment and a reactant,
+            # sum the fluxes accordingly (outward=negative, inward=positive)
+            if met in rxn.reactants and not_biomass_rxn and extracellular:
+                product_comps = set([p.compartment for p in rxn.products])
+                # if the reaction has no product (outward flux)
+                if len(product_comps) == 0:
+                    flux_sum[met] += -float(
+                        tra_bou_bio_fluxes[get_ids([rxn])])
+                # if the reaction has a product in "c" (inward flux)
+                elif main_comp in product_comps:
+                    flux_sum[met] += float(
+                        tra_bou_bio_fluxes[get_ids([rxn])])
+            # if the metabolite is in the "e" compartment and a product,
+            # sum the fluxes accordingly (outward=negative, inward=positive)
+            elif met in rxn.products and not_biomass_rxn and extracellular:
+                reactant_comps = set([p.compartment for p in rxn.reactants])
+                # if the reaction has no reactant (inward flux)
+                if len(reactant_comps) == 0:
+                    flux_sum[met] += float(
+                        tra_bou_bio_fluxes[get_ids([rxn])])
+                # if the reaction has a reactant in "c" (outward flux)
+                elif main_comp in reactant_comps:
+                    flux_sum[met] += -float(
+                        tra_bou_bio_fluxes[get_ids([rxn])])
+            # for all non-e metabolites (cannot be false positives, flux is
+            # automatically non-negative)
+            elif not_biomass_rxn and not extracellular:
+                flux_sum[met] += np.abs(
+                    float(tra_bou_bio_fluxes[get_ids([rxn])]))
+
+    return [met for met in flux_sum.T.loc[flux_sum.T['sum'] > 0].index]
```

Comment:
Albeit nicely commented, this function is really quite long. Let's take a look tomorrow whether we can break it down into smaller logical units.